doc/byte_stream.md

```diff
@@ -0,0 +1,137 @@
+***WARNING*** this is currently a proposal that is not implemented yet.
+
+This document describes a mechanism by which OpenMRN components can exchange a
+unidirectional untyped data stream. The API is laid out for compatibility with
+the StateFlow concept and provides asynchronous implementations to limit memory
+usage and flow control.
+
+## Use-cases
+
+- OpenLCB Stream sending. The client component sends data to the stream
+  transmitter using Byte Streams.
+  
+- OpenLCB Stream receiving. The client component is receiving data from the
+  Stream service using Byte Streams.
+  
+- A fast implementation of a TCP Hub should use Byte Streams to represent the
+  TCP messages for proxying between ports.
+  
+## Non-goals
+
+- It is not a goal to use the Byte Stream API to write from a State Flow to an
+  fd. (Neither sockets, pipes, nor physical files.) Writing to an fd should be
+  done by the native StateFlow features like `write_repeated`.
+
+## Requirements
+
+- Memory allocation
+
+  - Allocation has to avoid memory fragmentation. While the expectation is that
+    larger blocks of memory are allocated in one go, these blocks should be
+    reused between different users of Byte Streams and at different times,
+    instead of returning to malloc/free.
+  
+  - Block allocation size should be 1 kbyte. This is small enough that even
+    32kbyte MCUs can use the codebase, but large enough to capture meaningful
+    TCP packet sizes.
+  
+- Use and copy
+
+  - The implementation should be zero-copy. Once some data is in a byte buffer,
+    that data should not need to be copied in order to forward it to other byte
+    buffers or byte streams.
+    
+  - Reference counting should be available when the payload needs to be used
+    in multiple places.
+    
+  - It should be possible to take a subset of a buffer and send it to a
+    different byte stream. (Routers will do this.)
+    
+  - A special zero-copy mechanism should be available when the source data is
+    already in memory or in flash. Sending these bytes into the flow should not
+    need block allocation and data copy.
+
+- Flow control
+
+  - The data source has to be blocked in memory allocation when the sink has
+    not yet consumed the data.
+    
+  - The amount of read-ahead should be configurable (i.e., how much memory does
+    the source fill in before it gets blocked on waiting for the sink).
+
+
+## Implementation
+
+We define two buffer types.
+
+- `using RawBuffer = Buffer<char[1024]>;`
+
+  This buffer holds the raw data. This buffer is reference counted, shared
+  between all sinks, and never entered into a `Q`. All sinks consider this
+  buffer as read-only.
+
+- `using ByteBuffer = Buffer<ByteChunk>;`
+
+  This buffer holds a reference to a RawBuffer, and start/end pointers within
+  that describe the exact range of bytes to use. This buffer is not shareable,
+  it can be entered into a queue, i.e., sent to a `StateFlowWithQueue` /
+  `FlowInterface<ByteBuffer>`.
```

Comment:
Can a ByteBuffer hold a subset of another ByteBuffer, or only RawBuffer?

Reply:
This is described in detail under Implementation > Zero-copy.
But specifically to your question, there is no reason for a second level indirection. When a ByteBuffer is cloned, it takes a new reference to the same RawBuffer (if there is a reference to begin with), and copies the void* and size values from the other ByteBuffer.